viewing_party/party.py

```diff
@@ -1,23 +1,152 @@
+            add_to_watched(user_data, movie_dict)
+            tracker_dict = movie_dict
+    if tracker_dict:
+        user_data["watchlist"].remove(tracker_dict)
+
+    return user_data
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data): 
+    sum = 0 
 
+    if len(user_data["watched"]) != 0:
+        for i in range(len(user_data["watched"])):
+            sum += user_data["watched"][i]["rating"]
+        average_rating = sum / len(user_data["watched"])
+        return average_rating
+    
+    else:
+        return 0.0
+
+def get_most_watched_genre(user_data):
+    genre_list = []
+    counter = 0
+
+    if len(user_data["watched"]) == 0:
+        return None
+    
+    for i in range(len(user_data["watched"])):
+        genre_list.append(user_data["watched"][i]["genre"]) 
+        for i in genre_list:
+            frequency = genre_list.count(i)
+            if (frequency > counter):
+                counter = frequency
+                genre = i
+
+    return genre
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_friends_movie_titles(user_data):
+    friends_movies = set()
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_movies.add(movie["title"])
+    
+    return friends_movies
+
+def get_unique_watched(user_data):
+    friends_movies = get_friends_movie_titles(user_data)
+    unique_movies = []
+    
+    for movie in user_data["watched"]:
+        if movie["title"] not in friends_movies:
+            unique_movies.append(movie)
+
+    return unique_movies        
+
+def get_friends_unique_watched(user_data):
+    friends_movie_titles = get_friends_movie_titles(user_data)
+    user_movie_titles = set()
+
+    for movie in user_data["watched"]:
+        user_movie_titles.add(movie["title"])
+
+    friends_unique = friends_movie_titles - user_movie_titles
+    friends_unique_movies = []
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] in friends_unique:
+                if movie not in friends_unique_movies:
+                    friends_unique_movies.append(movie)
+
+    return friends_unique_movies    
 
         
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+    friends_watched = get_friends_unique_watched(user_data)
+    # If user data has a subscriptions key
+    if "subscriptions" in user_data:
+        user_subs = user_data["subscriptions"]
+        recs = []
+
+        for movie in friends_watched:
+            if movie["host"] in user_subs:
+                recs.append(movie)
+
+        return recs
+    
+    else:
+        return []

```

Comment:
Your function logic is correct here. If you had time, I would encourage thinking about how to refactor this code.

I'm pretty enthusiastic about refactoring, so I'll describe my own process of refactoring here:

First, I notice three interesting lines:`recs = []`, `return recs`, and `return []`. These three lines stand out to me because they all feel kind of similar. I think about how `return []` is the same as `return recs` if `recs` is an empty array, which it starts off as!

Then, I'd just try to change small pieces of code. I'd replace the `return` in the `else` clause with `return recs`.

After that, seeing that the line `return recs` is repeated, I'd delete the `else`, move a `return recs` line outside of the conditional, and delete all other `return recs`

Then, I'd move my definition of `recs` (the line `recs = []`) to the beginning of the function, so it is on the same indentation as when we `return` it.

Consider this, which passes the tests:

```python
def get_available_recs(user_data):
    friends_watched = get_friends_unique_watched(user_data)
    recs = []

    if "subscriptions" in user_data:
        user_subs = user_data["subscriptions"]
        for movie in friends_watched:
            if movie["host"] in user_subs:
                recs.append(movie)

    return recs
```

This refactored code uses only code you two wrote, but by moving a variable definition, we can get rid of the whole `else` clause.

PS: Sorry if I'm too enthusiastic or nerdy about this and wrote too much, I only hope it's helpful!